service/qos/qos_common.cc

```diff
@@ -120,4 +125,65 @@ void service_level_options::init_effective_names(sstring& service_level_name) {
     };
 }
 
+service::query_state& qos_query_state() {
+    using namespace std::chrono_literals;
+    const auto t = 10s;
+    static timeout_config tc{ t, t, t, t, t, t, t };
+    static thread_local service::client_state cs(service::client_state::internal_tag{}, tc);
+    static thread_local service::query_state qs(cs, empty_service_permit());
+    return qs;
+};
+
+static service_level_options::timeout_type get_duration(const cql3::untyped_result_set_row&row, std::string_view col_name) {
+    auto dur_opt = row.get_opt<cql_duration>(col_name);
+    if (!dur_opt) {
+        return qos::service_level_options::unset_marker{};
+    }
+    return std::chrono::duration_cast<lowres_clock::duration>(std::chrono::nanoseconds(dur_opt->nanoseconds));
+};
+
+future<qos::service_levels_info> get_service_levels(cql3::query_processor& qp, std::string_view ks_name, std::string_view cf_name, db::consistency_level cl) {
+    sstring prepared_query = format("SELECT * FROM {}.{};", ks_name, cf_name);
```

Comment:
it used to be static, currently what is keeping it alive?

Reply:
Right, `execute_internal` takes the query string by reference and `get_service_levels` is not a coroutine, so there is no guarantee that it will be alive long enough.

@Jadw1 it's probably best to convert `get_service_levels` to a coroutine

Reply:
it's not static because now the query can be done to `system_distributed.service_levels` and `system.service_levels_v2`.

fixed